### PR TITLE
Shared region enhancements

### PR DESCRIPTION
**Pull request by feldergast**
Added the ability to use SharedRegions when not all instances know the size of the region.

Added SharedArray and SharedMap as new ways to share readonly data.

### CODE REVIEW

File: src/sst/core/testingframework/sst_unittest.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def get_test_runtime_sec(self):
 
     def run_sst(self, sdl_file, out_file, err_file=None, set_cwd=None, mpi_out_files="",
                 other_args="", num_ranks=None, num_threads=None, global_args=None,
-                timeout_sec=60):
+                timeout_sec=60, expected_rc=0):
```
Comment (allevin) [inline src/sst/core/testingframework/sst_unittest.py:237]: Need to update the args section of the method's docstring explaining expected_rc 
```
global_args (str): Global Arguments provided from test engine args
expected_rc (int): The expected return code from the SST run
```

Need to do a check on expected_rc to verify its a valid data type
```
        if not (isinstance(timeout_sec, (int, float)) and not isinstance(timeout_sec, bool)):
            raise ValueError("ERROR: Timeout_sec must be a postive int or a float")
        check_param_type("expected_rc", expected_rc, int)  <-- add this ine
```

**Review by allevin — CHANGES_REQUESTED** [review]
Minor changes for the run_sst() function in terms of documentation and error checking.

File: tests/testsuite_default_SharedObject.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        testsuitedir = self.get_testsuite_dir()
+        outdir = test_output_get_run_dir()
+
+        model_options = '--model-options="{0}"'.format(options)
+        print(model_options)
+
+        # Set the various file paths
+        sdlfile = "{0}/test_SharedObject.py".format(testsuitedir)
+        #reffile = "{0}/sharedobject_tests/refFiles/test_{1}.out".format(testsuitedir, testtype)
+        outfile = "{0}/test_SharedObject_{1}.out".format(outdir, testtype)
```
Comment (allevin) [inline tests/testsuite_default_SharedObject.py:151]: The test run is displaying output to the console.  We need to add the err file to the run_sst
```
errfile = "{0}/test_SharedObject_{1}.err".format(outdir, testtype)

self.run_sst(sdlfile, outfile, errfile, other_args=model_options, expected_rc = exp_rc)
```

**Review by allevin — CHANGES_REQUESTED** [review]

File: tests/testsuite_default_SharedObject.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        self.sharedobject_test_template("map_partial_doubleinit", 1, "--param=object_type:map --param=num_entities:12 --param=full_initialization:false --param=double_initialize:true")
+        
+#####
+
+    def sharedobject_test_template(self, testtype, exp_rc, options):
+        testsuitedir = self.get_testsuite_dir()
+        outdir = test_output_get_run_dir()
+
+        model_options = '--model-options="{0}"'.format(options)
+        print(model_options)
```
Comment (allevin) [inline tests/testsuite_default_SharedObject.py:146]: change

```print(model_options)``` to  ```log_debug(model_options)```  to the the run clean and only display the model_options during a debug run.

**Review by allevin — CHANGES_REQUESTED** [review]

File: src/sst/core/testingframework/sst_unittest.py (hunk 1 above)
Comment (feldergast) [reply src/sst/core/testingframework/sst_unittest.py:237]: Added

File: tests/testsuite_default_SharedObject.py (hunk 3 above)
Comment (feldergast) [reply tests/testsuite_default_SharedObject.py:146]: Print not needed at all, just leftover from debugging.  Removed.

File: src/sst/core/sharedRegion.h [hunk 4]
```diff
@@ -12,11 +12,15 @@
...
 #define SST_CORE_CORE_SHAREDREGION_H
 
+#include "sst/core/warnmacros.h"
 #include "sst/core/sst_types.h"
 
 #include <string>
 #include <vector>
 
+#if !SST_BUILDING_CORE
+#warning "SharedRegion and it's accompanying classes have been deprecated and will be removed in SST 12. Please use the new SharedObject classes found in sst/core/shared."
```
Comment (gvoskuilen) [inline src/sst/core/sharedRegion.h:22]: This is tiny but "it's" -> "its" :)

**Review by gvoskuilen — DISMISSED** [review]

**Review by gvoskuilen — APPROVED** [review]

**Review by allevin — APPROVED** [review]

**Pull request merged**